R/plotting.R

```diff
@@ -0,0 +1,282 @@
+#' @importFrom crunch description name
+#' @importFrom rlang !! sym :=
+#' @export
+autoplot.NumericVariable <- function(x, ...) {
+    v <- as.vector(x)
+    plot_df <- data_frame(!!sym(name(x)) := v)
+    binwidth <- round((max(plot_df) - min(v)) / 5, 0)
+    ggplot(plot_df, aes(x = !!sym(name(x)))) +
+        geom_histogram(binwidth = binwidth, fill = card_colors[1]) +
+        theme_crunch() +
+        labs(title = name(x),
+            subtitle = description(x))
+}
+
+#' @rdname autoplot
+#' @export
+autoplot.CategoricalVariable <- function(x, ...) plotCategorical(x, ...)
+
+#' @rdname autoplot
+#' @export
+autoplot.CategoricalArrayVariable <- function(x, ...) plotCategorical(x, ...)
+
+#' @rdname autoplot
+#' @export
+autoplot.MultipleResponseVariable <- function(x, ...) plotCategorical(x, ...)
+
+#' @importFrom crunch  alias datasetReference loadDataset
+#' @importFrom ggplot2 autoplot
+plotCategorical <- function(x, ...) {
+    ds <- loadDataset(datasetReference(x))
+    cube <- crtabs(paste0("~", alias(x)), ds)
+    autoplot(cube, ...)
+}
+
+plot_fun_lookup <- function(plot_dim, plot_type) {
+    plot_fun <- paste0("crunch_", plot_dim, "d_", plot_type, "_plot")
+    return(get(plot_fun))
+}
+
+#' Autoplot methods for Crunch Objects
+#' 
+#' Generates ggplot representations of CrunchVariables and CrunchCubes
+#' 
+#' The Crunch autoplot methods generate plots which are optimized for various crunch objects.
+#' This allows you to visualize the object without bringing it into memory. You can select
+#' between three families of plots which will attempt to accomodate the dimensionality of
+#' the plotted object. These plots can be further extended and customized with other ggplot methods. 
+#' 
+#' @param x a CrunchCube, or CrunchVariable
+#' @param plot_type One of `"dot"`, `"grid"`, or `"bar"` which indicates the plot family
+#' you would like to use. Higher dimensional plots add color coding or facets depending
+#' on the dimensionality of the data. 
+#' @param measure The measure you wish to plot. This will usually be `"count"`, the default
+#' but can also be `".unweighted_counts"` or any other measure stored in the cube.
+#' @param ... ignored
+#' 
+#' @export
+#' @name autoplot
+#' @importFrom rlang !! !!! .data sym syms
+#' @importFrom purrr map map_chr
+#' @importFrom dplyr mutate filter pull
+#' @importFrom ggplot2 ggtitle
+autoplot.CrunchCube <- function(x, 
+    plot_type = c("dot", "grid", "bar"), 
+    measure = "count") {
+    plot_type = match.arg(plot_type)
+    display_names <- map(x@dims, "references") %>% 
+        map_chr("name")
+    
+    if (length(measure) > 1) {
+        # TODO think about how plots can support more than one measure. 
+        # for instance measures could be the grouping variable for 2d dot plots
+        stop("Autoplot can only support one measure.", .call = FALSE)
+    }
+    measure <- sym(measure)
+    
+    # TODO do we want to make this optional to allow people to plot missing entries?
```

Comment:
I believe the consistent way to do this is to follow the `useNA` setting and the utilities we have for setting that on the cube prior to passing into this function. It's not really `autoplot` if it grows 20 arguments.